File: sol/solution/0001-0100/0025/Solution.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reverseKGroup(head *ListNode, k int) *ListNode {
    curr := head
    count := 0
    for curr != nil && count < k {
        curr = curr.Next
        count++
    }
    if count == k {
        reversedHead := reverseLinkedList(head, k)
        head.Next = reverseKGroup(curr, k)
        return reversedHead
    }
    return head
}
func reverseLinkedList(head *ListNode, k int) *ListNode {
    var prev *ListNode = nil
    curr := head
    for k > 0 && curr != nil {
        next := curr.Next
        curr.Next = prev
        prev = curr
        curr = next
        k--
    }
    return prev
}

File: sol/solution/0001-0100/0025/Solution.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function reverseKGroup(head: ListNode | null, k: number): ListNode | null {
    let curr: ListNode | null = head;
    let count: number = 0;
    while (curr !== null && count < k) {
        curr = curr.next;
        count++;
    }
    if (count === k) {
        const reversedHead: ListNode | null = reverseLinkedList(head, k);
        if (head !== null) {
            head.next = reverseKGroup(curr, k);
        }
        return reversedHead;
    }
    return head;
}
function reverseLinkedList(head: ListNode | null, k: number): ListNode | null {
    let prev: ListNode | null = null;
    let curr: ListNode | null = head;
    while (k > 0 && curr !== null) {
        const next: ListNode | null = curr.next;
        curr.next = prev;
        prev = curr;
        curr = next;
        k--;
    }
    return prev;
}